### PR TITLE
Add trust_remote_code argument

### PR DESCRIPTION
**Pull request by lhoestq**
Draft about adding `trust_remote_code` to `load_dataset`.

```python
ds = load_dataset(..., trust_remote_code=True)  # run remote code (current default)
```

It would default to `True` (current behavior) and in the next major release it will prompt the user to check the code before running it (we'll communicate on this before doing it of course).

```python
# in the future
ds = load_dataset(...)  # prompt the user to check the code before running it (future default)
ds = load_dataset(..., trust_remote_code=True)  # run remote code
ds = load_dataset(..., trust_remote_code=False)  # disallow remote code
```

Related to https://github.com/huggingface/datasets/issues/6400

Will do a separate PR to use the parquet export when possible

### CODE REVIEW

File: src/datasets/load.py [hunk 1]
```diff
@@ -88,6 +89,45 @@
...
 
 
+def _raise_timeout_error(signum, frame):
+    raise ValueError(
+        "Loading this dataset requires you to execute custom code contained in the dataset repository on your local "
+        "machine. Please set the option `trust_remote_code=True` to permit loading of this dataset."
+    )
+
+
+def resolve_trust_remote_code(trust_remote_code: Optional[bool], repo_id: str) -> bool:
```
Comment (mariosasko) [inline src/datasets/load.py:99]: It complicates the logic, but Transformers does the same, so it's okay. 

A comment to the `transformers` version would be nice (to make it clear why it's implemented as-is)

**Review by mariosasko — COMMENTED** [review]
Nice! Some nits:

File: src/datasets/load.py [hunk 2]
```diff
@@ -1887,6 +2120,7 @@ def load_dataset(
     streaming: bool = False,
     num_proc: Optional[int] = None,
     storage_options: Optional[Dict] = None,
+    trust_remote_code: bool = None,
```
Comment (albertvillanova) [inline src/datasets/load.py:2123]: I think we should raise a `FutureWarning` to warn users that the default behavior will change in the future if they do not explicitly pass `trust_remote_code`.

**Review by albertvillanova — COMMENTED** [review]
Thanks. Just a suggestion below.

File: src/datasets/load.py (hunk 2 above)
Comment (lhoestq) [reply src/datasets/load.py:2123]: good idea !

**Comment by lhoestq** [comment]
I added tests and docs @mariosasko @albertvillanova let le know what you think !

**Comment by lhoestq** [comment]
I switched to using `deepmind/code_contests` in examples in the docs to avoid having to pass trust_remote_code, and remove the DEFAULT naming stuff :)

**Review by mariosasko — APPROVED** [review]
LGTM, thanks!

**Pull request merged**